src/ccover/ccover_bmc.cpp

```diff
@@ -0,0 +1,234 @@
+  else if(mm=="tso")
+    memory_model=util_make_unique<memory_model_tsot>(ns);
+  else if(mm=="pso")
+    memory_model=util_make_unique<memory_model_psot>(ns);
+  else
+  {
+    error() << "Invalid memory model " << mm
+            << " -- use one of sc, tso, pso" << eom;
+    throw "invalid memory model";
+  }
+}
+
+void ccover_bmct::setup()
+{
+  get_memory_model();
+  symex.options=options;
+
+  {
+    const symbolt *init_symbol;
+    if(!ns.lookup(INITIALIZE_FUNCTION, init_symbol))
+      symex.language_mode=init_symbol->mode;
+  }
+
+  status() << "Starting Bounded Model Checking" << eom;
+
+  symex.last_source_location.make_nil();
+
+  setup_unwind();
+}
+
+void ccover_bmct::run(abstract_goto_modelt &goto_model)
+{
+  setup();
+
+  auto get_goto_function = [&goto_model](const irep_idt &id) ->
+    const goto_functionst::goto_functiont &
+  {
+    return goto_model.get_goto_function(id);
+  };
+
+  // perform symbolic execution
+  symex.symex_from_entry_point_of(get_goto_function, symex_symbol_table);
+
+  // Borrow a reference to the goto functions map. This reference, or
+  // iterators pointing into it, must not be stored by this function or its
+  // callees, as goto_model.get_goto_function (as used by symex)
+  // will have side-effects on it.
+  const goto_functionst &goto_functions =
+    goto_model.get_goto_functions();
+
+  // add a partial ordering, if required
+  if(equation.has_threads())
+  {
+    memory_model->set_message_handler(get_message_handler());
+    (*memory_model)(equation);
+  }
+
+  statistics() << "size of program expression: "
+               << equation.SSA_steps.size()
+               << " steps" << eom;
+
+  slice();
+
+  const optionst::value_listt criteria=
+    options.get_list_option("cover");
+
+  cover(goto_functions, criteria);
+}
+
+void ccover_bmct::slice()
+{
+  // any properties to check at all?
+  if(equation.has_threads())
+  {
+    // we should build a thread-aware SSA slicer
+    statistics() << "no slicing due to threads" << eom;
+  }
+  else
+  {
+    if(options.get_bool_option("slice-formula"))
+    {
+      ::slice(equation);
+      statistics() << "slicing removed "
+                   << equation.count_ignored_SSA_steps()
+                   << " assignments"<<eom;
+    }
+    else
+    {
+      if(options.get_list_option("cover").empty())
+      {
+        simple_slice(equation);
+        statistics() << "simple slicing removed "
+                     << equation.count_ignored_SSA_steps()
+                     << " assignments"<<eom;
+      }
+    }
+  }
+
+  statistics() << "Generated "
+               << symex.total_vccs<<" VCC(s), "
+               << symex.remaining_vccs
+               << " remaining after simplification" << eom;
+}
+
+void ccover_bmct::setup_unwind()
+{
+  const std::string &set=options.get_option("unwindset");
+  std::vector<std::string> unwindset_loops;
+  split_string(set, ',', unwindset_loops, true, true);
```

Comment:
Elsewhere a function called `parse_unwindset` exists and should be used, quite likely before this PR goes in.